src/iron.rs

```diff
@@ -87,11 +90,68 @@ impl Protocol {
         match *self {
             Protocol::Http => "http",
             #[cfg(feature = "ssl")]
-            Protocol::Https { .. } => "https"
+            Protocol::Https => "https"
         }
     }
 }
 
+/// Protocol Handlers create servers
+pub trait ProtocolHandler<TListener: NetworkListener> {
+    /// Return the name used for this protocol in a URI's scheme part.
+    fn name(&self) -> &'static str;
+
+    /// Returns the protocol this represents
+    fn protocol(&self) -> Protocol;
+
+    /// Returns server for this protocol
+    fn create_server(&self, sock_addr: SocketAddr) -> Result<Server<TListener>, Error>;
+}
+
+/// Default HTTP handler
+pub struct HttpProtocolHandler;
+
+impl ProtocolHandler<HttpListener> for HttpProtocolHandler {
+    fn name(&self) -> &'static str { "http" }
+    fn protocol(&self) -> Protocol { Protocol::Http }
+
+    fn create_server(&self, sock_addr: SocketAddr) -> Result<Server, Error>{
+        Server::http(sock_addr)
+   }
+}
+
+/// Default HTTPs handler
+#[cfg(feature = "ssl")]
+pub struct HttpsProtocolHandler<S: Ssl + Clone + Send> {
```

Comment:
I'd just give this a constructor that takes in the `S` and avoid the Option + `impl HttpsProtocolHandler<Openssl>` altogether.